crates/kitsune_p2p/proxy/src/proxy_url.rs

```diff
@@ -0,0 +1,195 @@
+//! Utilities for dealing with proxy urls.
+
+use crate::*;
+
+/// Utility for deailing with proxy urls.
+#[derive(Debug, Display, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Deref, AsRef)]
+#[display(fmt = "{}", full)]
+pub struct ProxyUrl {
+    #[deref]
+    #[as_ref]
+    full: url2::Url2,
+    base: url2::Url2,
+}
+
+impl ProxyUrl {
+    /// Create a new proxy url from a full url str.
+    pub fn from_full(full: &str) -> TransportResult<Self> {
+        macro_rules! err {
+            () => {
+                TransportError::from(format!("Invalid Proxy Url: {}", full))
+            };
+        }
+        let full = url2::try_url2!("{}", full).map_err(|_| err!())?;
+        let base_scheme = match full.path_segments() {
+            None => return Err(err!()),
+            Some(mut s) => match s.next() {
+                None => return Err(err!()),
+                Some(s) => s,
+            },
+        };
+        let mut base = url2::url2!("{}://", base_scheme);
+        {
+            let mut path = full.path_segments().ok_or_else(|| err!())?;
+            path.next();
+            let mut mode = 0;
+            loop {
+                let key = match path.next() {
+                    None => break,
+                    Some(key) => key,
+                };
+                if key == "--" {
+                    mode = 1;
```

Comment:
Do you anticipate more than 2 modes in the future? Can you put in some comments briefly describing the modes? Or even make them an enum so we can have a name for each variant with documentation?

Reply:
What would happen if there were multiple instances of `"--"` in the URL? Multiple path_segments?

Reply:
This is like the `--` in command line processing... once it switches to mode "1" all path segments are treated as though they are on the base url.

Yeah, maybe I could make the mode a boolean? It's actually something like `before_base_path_marker`.

Reply:
added docs on the struct, and went with `found_base_path_marker` boolean instead of mode